### PR TITLE
allow `range(start, stop; ...)`

### PR DESCRIPTION
**Pull request by JeffBezanson**
This was discussed in #25896, and can be added now that deprecations are removed. This probably could have been in 1.0, and so might be ok to put in a point release. Thoughts? Otherwise it can go in 1.1.

### CODE REVIEW

**Comment by fredrikekre** [comment]
Should be v1.1.

**Comment by StefanKarpinski** [comment]
Let's not commit a flagrant violation of SemVer in literally our first post-1.0 release. @JeffBezanson, this is why I suggested that a fairly quick 1.1 on the order of 1-2 months might be in order because there are a number of little features like this that were unblocked by the removal of deprecations and which were not implemented yet in 1.0 on account of the accelerated time frame of the release.

**Review by mbauman — APPROVED** [review]

File: base/range.jl [hunk 1]
```diff
@@ -76,6 +76,34 @@ julia> range(1, step=5, stop=100)
 range(start; length::Union{Integer,Nothing}=nothing, stop=nothing, step=nothing) =
```
Comment (mbauman) [inline base/range.jl]: Maybe remove the docstring from this method? Or simplify it significantly? The two are essentially identical and will just end up looking repetitive in the docs.

**Comment by jlperla** [comment]
Is this PR (or an equivalent interface) *almost certainly* going into the v1.1 release?  To the extent that it can be discussed wither `Compat.jl` might support it?

**Comment by stillyslalom** [comment]
Any chance of including `range(start, stop) = range(start, stop, length=50)` to mimic the default-length behavior of Matlab and Numpy? I'm happy to put together the PR and docs if the core devs are open to it. It's a convenience function, sure, but it isn't going to hurt anyone.

**Comment by jlperla** [comment]
> Any chance of including range(start, stop) = range(start, stop, length=50) to mimic the default-length behavior of Matlab and Numpy?

That one I never liked in matlab, personally. I think the worry with the new range is that my guess on the default behavior (if any) is that `range(start, stop) = start:stop` which gives an appropriately generic step size of 1... But since `start:stop` is already so clean, I am not sure defaults in `range` make sense

**Comment by StefanKarpinski** [comment]
Why in the world is the default length 50?! Why not 100? That seems like a better and less surprising default. Defaulting to 50 is just something someone liked once upon a time that has now become enshrined for no reason whatsoever in various systems that copied each other. Let’s break the cycle of an arbitrary default value that never made any sense in the first place.

**Comment by jlperla** [comment]
@StefanKarpinski is this new interface sufficiently baked into Julia plans that we can try to make a case that it can go into `Compat.jl`?

This was a strategy brought up by @simonbyrne (though I want to be clear that he didn't endorse it necessarily).

The idea is that if the new range interface is finalized, that we would add it to Compat.jl so that people appalled by the current range interface can opt-in with the Compat.jl interface, at which point it would eventually just smoothly transition to v1.1

And with no semver violations, or conflicting homegrown type piracy between code bases to get the new interface.

**Comment by martinholters** [comment]
Once it's in Base, I see no reason it couldn't be added to Compat.

**Comment by stillyslalom** [comment]
I don't particularly care about the exact default length, but it's not completely arbitrary: 50 points are enough to capture the behavior of most smooth functions for plotting purposes, and "Break the tyranny of the sensible default value!" isn't an actual argument against default values. But this is a distracting bikeshed, and the PR's already an improvement over the current syntax, so I'll shut up.

**Comment by StefanKarpinski** [comment]
Pre-Compating this seems reasonable. There’s no real debate about what we’re going to do here.

**Comment by jlperla** [comment]
@StefanKarpinski Does this need to be merged before we can add it to `Compat.jl` (cc: @Nosferican )  If so, any movement towards merging is appreciated.

**Comment by AzamatB** [comment]
> Why not 100? That seems like a better and less surprising default

@StefanKarpinski, well, this leads to the proposal: let's make 100 as default. One possible justification for this choice is that the currently `range(start, stop)` defaults to `start:stop`, which means that it will never be used, since `start:stop` is way cleaner syntax, so it will always be given preference to. Besides defaulting to `step=1` is as random as choosing length 100 or 50. But contrary to `step=1`, `length=100` can actually be useful (e.g. when making quick plot of some smooth function, where you don't care about spacing so long as they capture the smootheness)

**Comment by JeffBezanson** [comment]
I don't think a default step of 1 is *as random* as a default length of 100 or 50. Consecutive integers aren't exactly some crazy thing we just made up. I guess it's slightly arbitrary when the arguments are not integer-valued though.

**Comment by AzamatB** [comment]
@JeffBezanson Yes, it can be quite difficult to recognize when seeing in someone's code something like `range(1, 2.9)` that it actually just means `1.0:1.0:2.0`. And, in fact, one could question validity of rounding down the `stop`. Might as well round it up so that `range(1, 2.9)` will yield `1.0:1.0:3.0` or round it to the nearest integer. This seems to me another arbitrary choice, which can be avoided if `length=100` is used.

I guess my point is that the current default has also arbitrary choices baked into it that make it confusing when reading someone's code. Furthermore, it is already covered by more slicker `a:b` syntax, which brings it's utility down to `0`. By contrast, `length=100` is maybe only `ε` more arbitrary, but it actually can be useful.

**Comment by jlperla** [comment]
Is a default for `step` or `length` necessary at all?  I am not sure I ever see `range(a, b)` being useful in generic code, for example, where `a:b` wouldn't be clearer?

**Comment by JeffBezanson** [comment]
A default step of 1 clearly makes sense for `1:n`, and we just take it from there. However I could see the case for allowing this only for integers (or maaybe integer-valued floats as well).

We could also require that a `length` or `step` keyword argument be passed for `range(start, stop)`.

**Comment by martinholters** [comment]
I'd still prefer to require either `length` or `step` to be given, as @StefanKarpinski proposed. (At least for now; it would leave the option of providing a default in the future.) But if we do want to support `range(start, stop)`, I definitely prefer the current `step=1` default over an arbitrary default for `length`.

**Comment by martinholters** [comment]
AV error looks unrelated but interesting.

**Comment by martinholters** [comment]
Ah, I wasn't even aware we supported `range(1, stop=100)` (with a default `step` of `1`). Seems slightly inconsistent not to support `range(1, 100)` then. Deprecating the former at this point is a no-go. I think my preferred route forward would be to _not_ support `range(1, 100)` and schedule `range(1, stop=100)` for future deprecation. (I cannot imagine it to be in widespread use, I'd always prefer to write `1:100`, but I might be wrong there.)

**Comment by JeffBezanson** [comment]
Comments addressed.

File: base/range.jl [hunk 2]
```diff
@@ -71,11 +73,28 @@ julia> range(1, step=5, length=100)
 
 julia> range(1, step=5, stop=100)
 1:5:96
+
+julia> range(1, 100)
+1:100
```
Comment (martinholters) [inline base/range.jl]: This doctest fails with the latest changes.

**Review by martinholters — CHANGES_REQUESTED** [review]
The docstring gives the default `step=1` which is not entirely correct. I'm not sure it's worth documenting the different behavior of the one-arg and two-arg versions. Maybe we restrict the documentation to the behavior of the two-arg version (either `step` or `length` has to be specified) and only silently keep supporting the old one-arg-with-neither-given?

**Comment by JeffBezanson** [comment]
Aha: the reason the default step of `1` is useful is that it allows returning a `UnitRange`.

**Comment by martinholters** [comment]
Ah, yes. Although that means the default is not really 1:
```julia
julia> range(1, stop=5)
1:5

julia> range(1, stop=5, step=1)
1:1:5
```
I'd expect giving the default explicitly not making any difference at all.

Irrespective of the `step=1` default business—should we schedule the `stop`-as-keyword-arg version for future deprecation?

**Comment by JeffBezanson** [comment]
I guess we could change the documentation to say the default is a `UnitRange` when no step is specified. But, `1:5 == 1:1:5` is true so this is getting rather pedantic.

I don't think `range(a, b)` and `range(a, stop=b)` are harmful and I don't see the need to deprecate anything here. In the current state of the PR `range(a, b)` requires an additional keyword argument, so we have the property that every call to `range` must have at least one keyword argument, which guarantees that every call site is pretty clear in its meaning.

**Pull request merged**